### ISSUE TITLE
Warning message for Quick-TRANSfer

### ISSUE THREAD

**Issue by QqQss** [immunogenomics/harmony#25]
When I run 'RunHarmony' for a seurat object, it will return a Warning message: 

> Quick-TRANSfer stage steps exceeded maximum (= 6687600)

What is this going? Does this will affect the final result?

**Comment by QqQss** [comment]
I find a solution: 
[https://stackoverflow.com/questions/21382681/kmeans-quick-transfer-stage-steps-exceeded-maximum]
This warning may be caused by kmean algorithms.
So could you tried to modified the default algorithms?

**Comment by ilyakorsunsky** [comment]
Thanks for finding this @QqQss! I will modify the code to allow you to try more iterations.

**Comment by ersgupta** [comment]
Curious, is this issue resolved? Also, do these warnings have any effect on the final result?

**Comment by kaizen89** [comment]
@ilyakorsunsky I am still getting this warning, is this normal? Is the output affected?
```
Harmony converged after 3 iterations
Warning messages:
1: Quick-TRANSfer stage steps exceeded maximum (= 7022450) 
```
Thanks

**Comment by ilyakorsunsky** [comment]
Dear @QqQss, @ersgupta, and @kaizen89, 

I modified the `RunHarmony` and `HarmonyMatrix` functions to allow you to specify the number of random restarts and number of iterations for the kmeans initialization. If you get the `Quick-TRANSfer stage steps exceeded maximum` warning, you can try to increase the number of maximum iterations and restarts: 

```{r}
obj <- RunHarmony(obj, 'dataset', kmeans_init_nstart=20, kmeans_init_iter_max=100)
```

Hope that helps! 

Best, 
Ilya

**Comment by madisonlayfield** [comment]
I am having this same error. I am using the HarmonyMatrix function and I tried to add your suggestion to my line of code, but I am getting an "unused argument" error for (kmeans_init_nstart=20, kmeans_init_iter_max=100).

Thanks.

**Comment by amjass12** [comment]
One question i have is, harmony continues to run and also converges even when this warning is present. can this warning therefore be ignored? thanks

**Comment by yingyonghui** [comment]
Hi @ilyakorsunsky ,
Thanks for your excellent tool.
I just came up with the same warning message and I checked the parameters by "?RunHarmony". I'm wondering whether you mean to adjust the argument "kmeans_init_iter_max" or "max.iter.cluster", or even "max.iter.harmony"?

Best,
Yingyong

**Comment by mcap91** [comment]
I just did some troubleshooting following all of these posts and it turns out my error was coming from samples with `Idents() `set to `NA`. I was using this NA containing variable to integrate with harmony. Once I set the NA idents to something (I used "unknown" because the cells were not labeled), this warning message disappeared.

**Comment by matasV99** [comment]
Hi, 

After modifying the code like this: "obj <- RunHarmony(obj, 'dataset', kmeans_init_nstart=20, kmeans_init_iter_max=100)"

I get a very weird convergence plot. Any idea what this means?
![Untitled (13)](https://user-images.githubusercontent.com/70501272/192959774-1852e8c8-5062-4162-9f0d-2a67d8271cb8.png)


Thank you,
Matas

**Comment by NadineBestard** [comment]
When I tried to add kmeans_init_nstart argument I got the same problem as @madisonlayfield https://github.com/immunogenomics/harmony/issues/25#issuecomment-612697181. Creating a Seurat object from my matrix, metadata and PCA to be able to use `RunHarmony()` instead worked. Looks like @ilyakorsunsky forgot to update `HarmonyMatrix()`
PS: I was working on the devel version of the package, updating that is the first thing I tried. Not sure if that's also a requirement for RunHarmony() to work, as the documentation is not updated.

**Comment by cenk-celik** [comment]
`gc(full = T)` before `RunHarmony()` worked for me.

**Comment by umutcakir** [comment]
I received the same warning message. I tried to set kmeans_init_iter_max and kmeans_init_nstart, and I used gc() command, but it still receive the same warning. Does this warning affect the result?

**Comment by matasV99** [comment]
Hi @umutcakir @NadineBestard and others,

I revisited this problem after some time off. I am not a developer of this, so this is just my limited understanding of the problem.

Tl;dr:
I did two things: a) increase the number of highly variable features I calculated my PCA latent space on and b) increased the number of random starts for the kmeans algorithm. 

Longer answer:
"1: Quick-TRANSfer stage steps exceeded maximum (= 9575900) " error means that "In rare cases, when some of the points (rows of ‘x’) are extremely close, the algorithm may not converge in the “Quick-Transfer” stage" (source: https://stackoverflow.com/questions/21382681/kmeans-quick-transfer-stage-steps-exceeded-maximum). My interpretation of this is that the latent PCA space has the points "too close to each other". In my case, I was running PCA in seurat with highly variable features that did not separate my data well enough (I only used 3000 and I had 150k cells). Once I ran PCA and Harmony with all features as input, harmony started converging. I still wanted to run Harmony with highly variable features in my dataset as that has been shown to improve batch correction and bio-signal preservation (source https://www.nature.com/articles/s41592-021-01336-8). Here is what functions I ran: "  merged <- RunPCA(object= merged,features = top4000_features, verbose = FALSE)" and "merged <- RunHarmony(merged, assay.use="SCT", group.by.vars = "model", kmeans_init_nstart=100, kmeans_init_iter_max=5000, plot_convergence = TRUE, verbose = TRUE, epsilon.cluster = -Inf, epsilon.harmony = -Inf, max.iter.harmony = 50)".

**Comment by pati-ni** [comment]
https://github.com/immunogenomics/harmony/issues/25#issuecomment-1261843791

If can add my 2 cents here, the warnings most likely are unrelated with the convergence plot. The warnings are emitted  by the kmeans from the R user space. 

For the convergence, I would suggest to use the latest version of harmony which mitigates these types of situations.

To the warnings. This is the initialization procedure of the cluster centroids. It is known that kmeans initial centroids positions might affect the convergence but it is hard to predict the outcome as this will be very data dependent. Having said that, this is only the initial position of the centroids and harmony moves those substantially during the main harmony routine.

We are considering more robust and faster initialization methods for this step which would remove the kmeans and the warnings but the solution is under active development.

Hopefully this will address some of the concerns regarding the warnings.

**Comment by me-orlov** [comment]
Hello everyone and thank you for this discussion!

Just wanted to chime in and say that I am also having the issue with weird convergence plots. I have downloaded the latest developer version of Harmony, but the convergence issues sadly persist in my case. Is there any consensus on whether it is acceptable to use the data associated with convergence plots like the one above? Can anything else be done to mitigate?

Thank you in advance!

**Comment by pati-ni** [comment]
Hi @me-orlov,

What do you mean developer version? The devel branch is defunct upstream. Can you confirm the actual loaded version in your sessionInfo()?

**Comment by pati-ni** [comment]
@me-orlov 

Also, please post the convergence plot

**Comment by me-orlov** [comment]
Hello pati-ni! My apologies for the late reply - I only now saw this. However, as the issue is still relevant, I would greatly appreciate it if you could help me out. The version of Harmony that I am running, according to SessionInfo(), is harmony_1.2.0. 

My plot looks like this. I am applying Harmony to spatial data.

![Screenshot_2024-01-16_17-21-03](https://github.com/immunogenomics/harmony/assets/119893114/2cc21003-14dc-4a5c-9015-35f433a053c1)

**Comment by pati-ni** [comment]
@me-orlov we are still in the process of solving this type of behavior. One thing you could try is to set lambda to NULL so it does some parameter tuning during runtime.

**Comment by yingyonghui** [comment]
Hi all, is there any update on this issue?
Any suggestions would be appreciated!

Thanks!

**Comment by pati-ni** [comment]
Hi @yingyonghui 

Please try out setting `lambda=NULL`. We have more features coming up to deal with this behavior properly.

**Comment by leonfodoulian** [comment]
Hi @pati-ni 

> Hi @yingyonghui
> 
> Please try out setting `lambda=NULL`. We have more features coming up to deal with this behavior properly.

Could you please provide more information regarding when the new features will be implemented and released?

Best,
Leon

**Comment by AmelZulji** [comment]
Hi @pati-ni, 

sorry to bother on this. All of the suggestions above didnt work for me, and I cant obtain stable results no matter what i try. I need to make a decision and would be great if you can provide an estimate when the new features will be implemented. 

thank you and wish you a nice day

**Comment by shanshenbing** [comment]
get similar warnings,and my commond is:

          `sr <- RunHarmony(sr, group.by.vars=c("patient_id", "orig.ident"),
                 plot_convergence=T, 
                 assay.use="RNA", reduction = "pca", dims.use=1:50,
                 theta=6, 
                 max.iter.harmony=20, max.iter.cluster=100,
                 #kmeans.algorithm="Lloyd", 
                 kmeans.iter.max=1000, kmeans.nstart=20,
                 epsilon.cluster=-Inf, epsilon.harmony=-Inf,
                 early_stop=0,ncores=10,
                 kmeans_init_nstart=20, kmeans_init_iter_max=100,
                 )` 
I think maybe @matasV99 is right. The key question is that some cells in PCA space are too close. if these close cells are inter-cluster cells,  using more genes as HVG is the right direction. But if they are intra-cluster cells, kmenas parameters should be changed.

**Comment by pati-ni** [comment]
> Hi @pati-ni,
> 
> sorry to bother on this. All of the suggestions above didnt work for me, and I cant obtain stable results no matter what i try. I need to make a decision and would be great if you can provide an estimate when the new features will be implemented.
> 
> thank you and wish you a nice day

@AmelZulji Is this somehow related to the quick-transfer warning message? Please open new issues for general troubleshooting of your data

**Comment by pati-ni** [comment]
@shanshenbing kmeans is used only to seed the cluster centroids for the harmony algorithm. I am unsure why the warning shows up but it should not affect the algorithm. Unless someone suspects that the resulting cluster centroids are problematic this warning should not be a concern.

If you can't stand the warnings this unstable branch contains a custom initialization procedure (faster and no warnings): 

https://github.com/pati-ni/harmony/tree/census-integration

However, it is unstable due to upstream bugs in Rcpp in conda deployments https://github.com/RcppCore/RcppArmadillo/issues/438  

We will push upstream soon most likely with a work around

**Comment by AmelZulji** [comment]
Sorry for late response.
> @AmelZulji Is this somehow related to the quick-transfer warning message? Please open new issues for general troubleshooting of your data

yes, the results are not reproducible if the warning appears. The general biological signal is preserved, but clusters are slightly different and usually numbered differently which requires to readjust cluster annotation everytime the pipeline is run.

**Comment by eray-sahin** [comment]
Hello again,

I am having the same issue for higher resolutions in spatial trx data, and all the suggestions above, including changing and increasing kmeans_init_nstart and kmeans_init_iter_max did not help. Only remaining try, described [here](https://stackoverflow.com/questions/21382681/kmeans-quick-transfer-stage-steps-exceeded-maximum), seems to be changing the algorithm in kmeans function but I believe it is encoded in C/C++, so no chance to change it. Is there any plan to update it, or should we change our analysis strategy?
